### PR TITLE
feat: testcontainer setup

### PR DESCRIPTION
**Pull request by aritra1999**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new target in the Makefile to run tests.
	- Added a new PostgreSQL database connection feature.
	- Implemented a structured setup for PostgreSQL database containers for testing.
	- Created a new test suite for the main application, including setup and teardown functions.
	- Enhanced GitHub Actions workflow to create a `.env` file before build and test processes.

- **Bug Fixes**
	- Improved error handling for database connections and container management.

- **Documentation**
	- Updated comments and structure for clarity in the new test and setup files.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,2 +1,5 @@
 dev:
 	go run main.go
+
+test:
+	go test -v ./${module}...
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply Makefile]: Thank you for the suggestion, I'll update it in the next commit.

[the remark this answers is not in the document]
Comment (aritra1999) [reply Makefile]: Thank you for the suggestion, I'll update it in the next commit.

File: tests/setup.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+type TestDBContainer struct {
+	Id         string
+	ConnString string
+}
+
+func Setup(t *testing.T) (testcontainers.Container, error) {
+	return SetupPostgresContainer(t)
+}
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply tests/setup.go]: That's an awesome suggestion, I'll add it in the next commit!

File: tests/setup.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		ContainerRequest: req,
+		Started:          true,
+	})
+	if err != nil {
+		t.Fatalf("failed to start container: %s", err)
+		return nil, err
+	}
+
+	return postgresC, nil
+}
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply tests/setup.go]: I'll revisit this later!

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -5,24 +5,23 @@
...
 
-    - name: Build
-      run: go build -v ./...
+      - name: Build
+        run: touch .env && go build -v ./...
 
-    - name: Test
-      run: go test -v ./...
+      - name: Test
+        run: touch .env && go test -v ./...
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply .github/workflows/ci.yml:27]: We don't need these variables while testing, but still need an empty `.env` file because it's loaded in the main function.

[the remark this answers is not in the document]
Comment (aritra1999) [reply .github/workflows/ci.yml:27]: Will revisit this setup later

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -5,24 +5,23 @@ name: Go
...
+      - uses: actions/checkout@v4
 
-    - name: Set up Go
-      uses: actions/setup-go@v4
-      with:
-        go-version: '1.23'
+      - name: Set up Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: "1.23"
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply .github/workflows/ci.yml:21]: It's working with version 1.23

File: go.mod [hunk 6]
```diff
@@ -3,18 +3,64 @@
 go 1.23.1
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply go.mod:3]: It's working with version 1.23

File: go.mod [hunk 7]
```diff
@@ -3,18 +3,64 @@ module janitor
...
 	github.com/tursodatabase/libsql-client-go v0.0.0-20240902231107-85af5b9d094d // indirect
+	github.com/yusufpapurcu/wmi v1.2.3 // indirect
+	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp v0.49.0 // indirect
+	go.opentelemetry.io/otel v1.24.0 // indirect
+	go.opentelemetry.io/otel/metric v1.24.0 // indirect
+	go.opentelemetry.io/otel/trace v1.24.0 // indirect
+	golang.org/x/crypto v0.22.0 // indirect
 	golang.org/x/exp v0.0.0-20240325151524-a685a6edb6d8 // indirect
 	golang.org/x/sys v0.26.0 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply go.mod:65]: Will revisit this later

File: tests/setup.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	"time"
+
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+type TestDBContainer struct {
+	Id         string
+	ConnString string
+}
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply tests/setup.go]: Thank you for the suggestion, I'll update it in the next commit.

File: main_test.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	}
+
+	dbConnString, err := tests.GetConnString(t, postgresC)
+	db.ConnectDB(dbConnString)
+
+	g.Describe("getExpiredUptimeChecks", func() {
+		g.It("should return a list of expired uptime checks", func() {
+			g.Assert(1).Equal(1)
+		})
+	})
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply main_test.go:28]: This is a placeholder test for now, it'll be replaced with actual tests in the next pull request!

File: tests/teardown.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+	"github.com/testcontainers/testcontainers-go"
+)
+
+func Teardown(container *testcontainers.Container) {
+	ctx := context.Background()
+	if err := (*container).Terminate(ctx); err != nil {
+		log.Fatalf("failed to terminate container: %v", err)
+	}
+}
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply tests/teardown.go:15]: There's nothing we can do with the error where the teardown function is called, which is why we don't return anything from the teardown function.

File: db/db.go [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+func ConnectDB(connString string) (*sql.DB, error) {
+	log.Print("Connecting to database")
+	db, err := sql.Open("postgres", connString)
+	if err != nil {
+		return nil, err
+	}
+	return db, nil
+}
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply db/db.go:17]: This is an internal job that does not require the pool setup, it'll connect, process, and disconnect from the database after it's done processing.

File: main_test.go [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+	tests.Teardown(&postgresC)
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply main_test.go]: Thank you for the suggestion, I'll update it in the next commit.

File: main_test.go [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+	g := Goblin(t)
+
+	postgresC, err := tests.Setup(t)
+	if err != nil {
+		t.Fatalf("failed to setup postgres container: %s", err)
+		return
+	}
+
+	dbConnString, err := tests.GetConnString(t, postgresC)
+	db.ConnectDB(dbConnString)
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply main_test.go]: This had been updated!

File: tests/setup.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	port, err := postgresC.MappedPort(ctx, "5432")
+	if err != nil {
+		t.Fatalf("failed to get container port: %s", err)
+		return "", err
+	}
+
+	return fmt.Sprintf("host=%s port=%s user=test password=test dbname=testdb sslmode=disable", host, port.Port()), nil
+}
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply tests/setup.go:43]: Unnecessary

File: main_test.go [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+	db.ConnectDB(dbConnString)
+
+	g.Describe("getExpiredUptimeChecks", func() {
+		g.It("should return a list of expired uptime checks", func() {
+			g.Assert(1).Equal(1)
+		})
+	})
+
+	defer tests.Teardown(&postgresC)
```
[the remark this answers is not in the document]
Comment (aritra1999) [reply main_test.go:30]: It started throwing an error
